### PR TITLE
Pact Consumer

### PR DESCRIPTION
**Pull request by JP-Ellis**
## :memo: Summary

Implements the core `Pact` class, along with a number of other classes required for this to all work.

## ~:rotating_light: Breaking Changes~

No changes to the existing library, but introduces significant changes to `pact.v3`.

## :hammer: Test Plan

Unit tests have been included as part of this PR.

## :link: Related issues/PRs

- Resolves: #382

## :eyes: Notes for Reviewers

As this PR is rather substantial, I have split some of the less consequential commits from the major commit (`feat(v3): implement pact class`). While the PR as a whole can be reviewed, I would recommend reviewing each commit individually.

### CODE REVIEW

File: pact/v3/ffi.py [hunk 1]
```diff
@@ -16,26 +16,79 @@
...
 documentation](https://cffi.readthedocs.io/en/latest/using.html).
+
+### Handles
+
+The Rust library exposes a number of handles to internal data structures. This
+is done to avoid exposing the internal implementation details of the library to
+users of the library, and avoid unnecessarily casting to and from possibly
+complicated structs.
+
+In the Rust library, the handles are thin wrappers around integers, and
```
Comment (rholshausen) [inline pact/v3/ffi.py:40]: This was done to make it easier for Go and JS, but if it is a problem, we can change it to be a pointer type

**Review by rholshausen — APPROVED** [review]
LGTM

File: pact/v3/ffi.py (hunk 1 above)
Comment (JP-Ellis) [reply pact/v3/ffi.py:40]: It's not at all an issue. It's just that a `struct { int }` has the same in-memory representation as a bare `int`. I added the thin wrappers for two reasons:

- Obfuscating the `int`. After all, it doesn't make sense to 'add' or 'subtract' handles (unlike integers). It also makes it explicit that the handles to a Pact or an Interaction are different.
- This makes it easy to implement the `__del__` method to ensure that memory is cleaned up when Python's garbage control executes.

Comment (YOU54F) [reply pact/v3/ffi.py:40]: Great explanation @JP-Ellis 👍🏾 thank you!

**Review by YOU54F — APPROVED** [review]
awesome!!!

**Comment by YOU54F** [comment]
>🔗 Related issues/PRs
Resolves: https://github.com/pact-foundation/pact-python/issues/376

Should this be Add support and example of a HTTP consumer test without the plugin bit. I couldn't see a plugin being loaded in any of the unit tests.

The methods are exposed, but there isn't an example unit test as per linked ticket (that i've seen)

**Comment by YOU54F** [comment]
also added `g++` to the linux image deps to support a couple of python versions that were failing to build

**Comment by JP-Ellis** [comment]
> > 🔗 Related issues/PRs
> > Resolves: https://github.com/pact-foundation/pact-python/issues/376
> 
> Should this be Add support and example of a HTTP consumer test without the plugin bit. I couldn't see a plugin being loaded in any of the unit tests.
> 
> The methods are exposed, but there isn't an example unit test as per linked ticket (that i've seen)

I linked to the wrong underlying issue 😅. Fixed.

As with the related PRs, I had to rebase this before merging it.

**Pull request merged**